File: C++ Repository/Edge.cpp


#include "Edge.hpp"
#include "Vertex.hpp"
#include <sstream>
#include "DebugPrint.hpp"


Edge::Edge(Vertex *startVertex,Vertex *endVertex,Overlap *overlap,Read *read,int beginIdx,int endIdx,int type){
    this->startVertex=startVertex;
    this->endVertex=endVertex;
    this->read=read;
    read->REFERENCE++;
    this->overlap=overlap;
    this->beginIdx=beginIdx;
    this->endIdx=endIdx;
    this->type=type;
    startVertex->getOutEdges().push_back(this);
    endVertex->getInEdges().push_back(this);
    startVertex->sortEdges();
}

std::string Edge::toDNA(){
    std::string DNA=read->get(beginIdx, endIdx);
    for(Edge *e:containedEdges){
        DNA+=e->toDNA();
    }
    return DNA;
}

std::string Edge::toGFA(){
    std::stringstream ss;
    ss << "L\t" <<
				overlap->leftPart()->getRead()->getID() <<
				"\t" << (overlap->leftPart()->isForward()?"+":"-") << "\t" <<
				overlap->rightPart()->getRead()->getID() <<
				"\t" << (overlap->rightPart()->isForward()?"+":"-") << "\t" <<
				overlap->size() << "M";
    return ss.str();
}

std::string Edge::toGFARecursive(){
    std::string GFA=toGFA();
    for(Edge *e:containedEdges){
        GFA.append("\n");
        GFA.append(e->toGFARecursive());
    }
    return GFA;
}


std::string Edge::toString() {
    std::stringstream ss;
    ss << startVertex->toString() << "->" << endVertex->toString() << " TYPE:" << type << " LEN:" << length() << " CONT:"<< containedEdges.size();
    return ss.str();
}

std::vector<Overlap*> Edge::getOverlaps(){
    std::vector<Overlap*> overlaps;
    overlaps.push_back(getOverlap());
    for(Edge *edge:containedEdges){
        std::vector<Overlap*> eo=edge->getOverlaps();
        overlaps.insert(overlaps.begin(),eo.begin(),eo.end());
    }
    return overlaps;
}

Overlap* Edge::getOverlap(){
    return overlap;
}

bool Edge::hasContainedEdges(){
    return containedEdges.size()>0;
}

int Edge::hashCode() {
    return 31*startVertex->hashCode()+37*endVertex->hashCode();
}

Vertex* Edge::getStartVertex(){
    return startVertex;
}

Read* Edge::getRead(){
    return read;
}

Vertex* Edge::getEndVertex(){
    return endVertex;
}

int Edge::segmentLength(){
    int l=beginIdx-endIdx;
    if(l<0){
        l=-l;
    }
    return l;
}

int Edge::numContainedEdges(){
    int num=0;
    for(Edge *edge:containedEdges){
        num+=1;
        num+=edge->numContainedEdges();
    }
    return num;
}

Edge::~Edge(){
    for(Edge *edge:containedEdges){
        delete edge;
    }
    overlap->REFERENCE--;
    if(overlap->REFERENCE==0){
        delete overlap;   
    }
    read->REFERENCE--;
    if(read->REFERENCE==0){
        delete read;
    }
}

int Edge::length(){
    int length=segmentLength();
    for(Edge *edge:containedEdges){
        length+=edge->length();
    }
    return length;
}

void Edge::remove(){
    startVertex->getOutEdges().erase(std::find(startVertex->getOutEdges().begin(), startVertex->getOutEdges().end(), this));
    endVertex->getInEdges().erase(std::find(endVertex->getInEdges().begin(), endVertex->getInEdges().end(), this));
}

void Edge::addEdge(Edge *edge){
    containedEdges.push_back(edge);
    endVertex->getInEdges().erase(std::find(endVertex->getInEdges().begin(), endVertex->getInEdges().end(), this));
    endVertex=edge->getEndVertex();
    endVertex->getInEdges().push_back(this);
    edge->remove();
}


bool Edge::hasInteriorVertices() {
    return containedEdges.size()>0;
}

File: Java repository/StringGraph_layout/src/graph/Edge.java
package graph;

import java.util.ArrayList;

import overlap.Overlap;
import overlap.Read;

//class for representing edges

public class Edge {

	private Vertex startVertex;
	private Vertex endVertex;
	private Read read;
	private Overlap overlap;
	private int beginIdx;
	private int endIdx;
	public boolean reduce;
	public final int type;
	private ArrayList<Edge> containedEdges=new ArrayList<>();
	
	public String toDNA(){
		//System.out.println(read);
		//System.out.println(overlap);
		String DNA=read.get(beginIdx, endIdx);
		for(Edge e:containedEdges){
			DNA+=e.toDNA();
		}
		return DNA;
	}
	
	public String toGFARecursive(){
		String GFA=toGFA();
		for(Edge e:containedEdges){
			GFA=GFA+"\n";
			GFA=GFA+e.toGFARecursive();
		}
		return GFA;
	}

	
	public String toGFA(){
		return
				"L\t"+
				overlap.leftPart().getRead().getID()+
				"\t"+(overlap.leftPart().isForward()?"+":"-")+"\t"+
				overlap.rightPart().getRead().getID()+
				"\t"+(overlap.rightPart().isForward()?"+":"-")+"\t"+
				overlap.size()+"M";
	}
	
	@Override
	public String toString() {
		return startVertex+"->"+endVertex+" TYPE:"+type+" LEN:"+length()+" CONT:"+containedEdges;//"Len:"+length();
	//	return overlap.IDX+"";//+" LEN:"+length();
	}
	
	public ArrayList<Overlap> getOverlaps(){
		ArrayList<Overlap> overlaps=new ArrayList<>();
		overlaps.add(getOverlap());
		for(Edge edge:containedEdges){
			overlaps.addAll(edge.getOverlaps());
		}
		return overlaps;
	}
	
	public Overlap getOverlap(){
		return overlap;
	}
	
	public boolean hasContainedEdges(){
		return containedEdges.size()>0;
	}
	
	@Override
	public int hashCode() {
		return 31*startVertex.hashCode()+37*endVertex.hashCode();
	}
	
	public Vertex getStartVertex(){
		return startVertex;
	}
	
	public Read getRead(){
		return read;
	}

	public Vertex getEndVertex(){
		return endVertex;
	}
	
	public int maxLength(){
		int max=segmentLength();
		for(Edge e:containedEdges){
			if(e.maxLength()>max){
				max=e.maxLength();
			}
		}
		return max;
	}

	public double minLo(){
		double min=1.0*overlap.size()/read.length();
		for(Edge e:containedEdges){
			if(e.minLo()<min){
				min=e.minLo();
			}
		}
		//System.out.println(min);
		return min;
	}

	public double maxLo(){
		double max=1.0*overlap.size()/read.length();
		for(Edge e:containedEdges){
			if(e.maxLo()>max){
				max=e.maxLo();
			}
		}
		//System.out.println(max);
		return max;
	}
	
	public int maxSegmentLength(){
		int max=segmentLength();
		for(Edge e:containedEdges){
			if(e.maxSegmentLength()>max){
				max=e.maxSegmentLength();
			}
		}
		return max;
	}
	
	public int minSegmentLength(){
		int min=segmentLength();
		for(Edge e:containedEdges){
			if(e.minSegmentLength()<min){
				min=e.minSegmentLength();
			}
		}
		return min;
	}



	public int segmentLength(){
		int l=beginIdx-endIdx;
		if(l<0){
			l=-l;
		}
		return l;
	}

	public int numContainedEdges(){
		int num=0;
		for(Edge edge:containedEdges){
			num+=1;
			num+=edge.numContainedEdges();
		}
		return num;
	}
	
	public int length(){
		int length=segmentLength();
		for(Edge edge:containedEdges){
			length+=edge.length();
		}
		return length;
	}
	
	public void remove(){
		startVertex.getOutEdges().remove(this);
		endVertex.getInEdges().remove(this);
	}
	
	public void addEdge(Edge edge){
		containedEdges.add(edge);
		endVertex.getInEdges().remove(this);
		endVertex=edge.getEndVertex();
		endVertex.getInEdges().add(this);
		edge.remove();
	}
	
	public Edge(Vertex startVertex,Vertex endVertex,Overlap overlap,Read read,int beginIdx,int endIdx,int type){
		this.startVertex=startVertex;
		this.endVertex=endVertex;
		this.read=read;
		this.overlap=overlap;
		this.beginIdx=beginIdx;
		this.endIdx=endIdx;
		this.type=type;
		startVertex.getOutEdges().add(this);
		endVertex.getInEdges().add(this);
		startVertex.sortEdges();
		//System.out.println(this);
	}

	public boolean hasInteriorVertices() {
		return containedEdges.size()>0;
	}
	
}
